src/util/dirty.cpp

```diff
@@ -0,0 +1,382 @@
+    }
+
+    SPDLOG_TRACE("Set up dirty tracking signal handler");
+}
+
+void SegfaultDirtyTracker::handler(int sig,
+                                   siginfo_t* info,
+                                   void* ucontext) noexcept
+{
+    void* faultAddr = info->si_addr;
+
+    if (!tracking.isInitialised()) {
+        // Unexpected segfault, treat as normal
+        faabric::util::handleCrash(sig);
+    }
+
+    tracking.markDirtyPage(faultAddr);
+
+    // Align down to nearest page boundary
+    uintptr_t addr = (uintptr_t)faultAddr;
+    addr &= -HOST_PAGE_SIZE;
+    auto* alignedAddr = (void*)addr;
+
+    // Remove write protection from page
+    if (::mprotect(alignedAddr, HOST_PAGE_SIZE, PROT_READ | PROT_WRITE) == -1) {
+        SPDLOG_ERROR("WARNING: mprotect failed to unset read-only");
+    }
+}
+
+void SegfaultDirtyTracker::clearAll()
+{
+    tracking = ThreadTrackingData();
+}
+
+void SegfaultDirtyTracker::startThreadLocalTracking(std::span<uint8_t> region)
+{
+    if (region.empty() || region.data() == nullptr) {
+        return;
+    }
+
+    SPDLOG_TRACE("Starting thread-local tracking on region size {}",
+                 region.size());
+    tracking = ThreadTrackingData(region);
+}
+
+void SegfaultDirtyTracker::startTracking(std::span<uint8_t> region)
+{
+    SPDLOG_TRACE("Starting tracking on region size {}", region.size());
+
+    if (region.empty() || region.data() == nullptr) {
+        return;
+    }
+
+    PROF_START(MprotectStart)
+
+    // Note that here we want to mark the memory read-only, this is to ensure
+    // that only writes are counted as dirtying a page.
+    if (::mprotect(region.data(), region.size(), PROT_READ) == -1) {
+        SPDLOG_ERROR("Failed to start tracking with mprotect: {} ({})",
+                     errno,
+                     strerror(errno));
+        throw std::runtime_error("Failed mprotect to start tracking");
+    }
+
+    PROF_END(MprotectStart)
+}
+
+void SegfaultDirtyTracker::stopTracking(std::span<uint8_t> region)
+{
+    if (region.empty() || region.data() == nullptr) {
+        return;
+    }
+
+    SPDLOG_TRACE("Stopping tracking on region size {}", region.size());
+
+    PROF_START(MprotectEnd)
+
+    if (::mprotect(region.data(), region.size(), PROT_READ | PROT_WRITE) ==
+        -1) {
+        SPDLOG_ERROR("Failed to stop tracking with mprotect: {} ({})",
+                     errno,
+                     strerror(errno));
+        throw std::runtime_error("Failed mprotect to stop tracking");
+    }
+
+    PROF_END(MprotectEnd)
+}
+
+void SegfaultDirtyTracker::stopThreadLocalTracking(std::span<uint8_t> region)
+{
+    // Do nothing - need to preserve thread-local data for getting dirty regions
+    SPDLOG_TRACE("Stopping thread-local tracking on region size {}",
+                 region.size());
+}
+
+void SegfaultDirtyTracker::reinitialise()
+{
+    if (faabric::util::isTestMode()) {
+        // This is a hack because catch changes the segfault signal handler
```

Comment:
This is easily configurable with catch's defines: <https://github.com/catchorg/Catch2/blob/devel/docs/configuration.md#other-toggles>, specifically `CATCH_CONFIG_NO_POSIX_SIGNALS` can be defined

Reply:
Ah awesome, did not know about that.